azure/services/agentpools/spec.go

```diff
@@ -0,0 +1,202 @@
+	MaxCount *int32 `json:"maxCount,omitempty"`
+
+	// Minimum number of nodes for auto-scaling
+	MinCount *int32 `json:"minCount,omitempty"`
+
+	// Node labels - labels for all of the nodes present in node pool
+	NodeLabels map[string]*string `json:"nodeLabels,omitempty"`
+
+	// NodeTaints specifies the taints for nodes present in this agent pool.
+	NodeTaints []string `json:"nodeTaints,omitempty"`
+
+	// EnableAutoScaling - Whether to enable auto-scaler
+	EnableAutoScaling *bool `json:"enableAutoScaling,omitempty"`
+
+	// AvailabilityZones represents the Availability zones for nodes in the AgentPool.
+	AvailabilityZones []string
+
+	// MaxPods specifies the kubelet --max-pods configuration for the agent pool.
+	MaxPods *int32 `json:"maxPods,omitempty"`
+
+	// OsDiskType specifies the OS disk type for each node in the pool. Allowed values are 'Ephemeral' and 'Managed'.
+	OsDiskType *string `json:"osDiskType,omitempty"`
+
+	// EnableUltraSSD enables the storage type UltraSSD_LRS for the agent pool.
+	EnableUltraSSD *bool `json:"enableUltraSSD,omitempty"`
+
+	// OSType specifies the operating system for the node pool. Allowed values are 'Linux' and 'Windows'
+	OSType *string `json:"osType,omitempty"`
+
+	// Headers is the list of headers to add to the HTTP requests to update this resource.
+	Headers map[string]string
+}
+
+// ResourceName returns the name of the agent pool.
+func (s *AgentPoolSpec) ResourceName() string {
+	return s.Name
+}
+
+// ResourceGroupName returns the name of the resource group.
+func (s *AgentPoolSpec) ResourceGroupName() string {
+	return s.ResourceGroup
+}
+
+// OwnerResourceName is a no-op for agent pools.
+func (s *AgentPoolSpec) OwnerResourceName() string {
+	return s.Cluster
+}
+
+// CustomHeaders returns custom headers to be added to the Azure API calls.
+func (s *AgentPoolSpec) CustomHeaders() map[string]string {
+	return s.Headers
+}
+
+// Parameters returns the parameters for the agent pool.
+func (s *AgentPoolSpec) Parameters(existing interface{}) (params interface{}, err error) {
+	if existing != nil {
+		existingPool, ok := existing.(containerservice.AgentPool)
+		if !ok {
+			return nil, errors.Errorf("%T is not a containerservice.AgentPool", existing)
+		}
+
+		// agent pool already exists
+		ps := *existingPool.ManagedClusterAgentPoolProfileProperties.ProvisioningState
+		if ps != string(infrav1.Canceled) && ps != string(infrav1.Failed) && ps != string(infrav1.Succeeded) {
+			msg := fmt.Sprintf("Unable to update existing agent pool in non terminal state. Agent pool must be in one of the following provisioning states: Canceled, Failed, or Succeeded. Actual state: %s", ps)
+			return nil, azure.WithTransientError(errors.New(msg), 20*time.Second)
+		}
+
+		// Normalize individual agent pools to diff in case we need to update
+		existingProfile := containerservice.AgentPool{
+			ManagedClusterAgentPoolProfileProperties: &containerservice.ManagedClusterAgentPoolProfileProperties{
```

Comment:
See #2559 which added labels to the normalized diff

Reply:
I've rebased and resolved the conflicts!